upload-api/test/service/ucan-invocation.test.js

```diff
@@ -0,0 +1,138 @@
+
+test.before(async t => {
+  const { client: s3Client, clientOpts: s3ClientOpts } = await createS3({ port: 9000 })
+
+  t.context.s3Client = s3Client
+  t.context.s3ClientOpts = s3ClientOpts
+})
+
+test('parses ucan invocation request', async t => {
+  const uploadService = await Signer.generate()
+  const alice = await Signer.generate()
+  const { proof, spaceDid } = await createSpace(alice)
+
+  const data = new Uint8Array([11, 22, 34, 44, 55])
+  const link = await CAR.codec.link(data)
+  const nb = { link, size: data.byteLength }
+  const can = 'store/add'
+
+  const request = await CAR.encode([
+    {
+      issuer: alice,
+      audience: uploadService,
+      capabilities: [{
+        can,
+        with: spaceDid,
+        nb
+      }],
+      proofs: [proof],
+    }
+  ])
+
+  // @ts-expect-error different type interface in AWS expected request
+  const ucanInvocationObject = await parseUcanInvocationRequest(request)
+
+  const requestCar = await CAR.codec.decode(request.body)
+  const requestCarRootCid = requestCar.roots[0].cid
+
+  t.is(ucanInvocationObject.carCid, requestCarRootCid.toString())
+  t.truthy(ucanInvocationObject.bytes)
+
+  // Decode and validate bytes
+  const ucanCar = await CAR.codec.decode(ucanInvocationObject.bytes)
+  // @ts-expect-error UCAN.View<UCAN.Capabilities> inferred as UCAN.View<unknown>
+  const ucan = UCAN.decode(ucanCar.roots[0].bytes)
+
+  t.is(ucan.iss.did(), alice.did())
+  t.is(ucan.aud.did(), uploadService.did())
+  t.deepEqual(ucan.prf, [proof.root.cid])  
+  t.is(ucan.att.length, 1)
+  t.like(ucan.att[0], {
+    nb,
+    can,
+    with: spaceDid,
+  })
+})
+
+test('persists ucan invocation CAR file', async t => {
+  const { bucketName } = await prepareResources(t.context.s3Client)
+  const ucanStore = createUcanStore('us-west-2', bucketName, t.context.s3ClientOpts)
+
+  const uploadService = await Signer.generate()
+  const alice = await Signer.generate()
+  const { proof, spaceDid } = await createSpace(alice)
+
+  const data = new Uint8Array([11, 22, 34, 44, 55])
+  const link = await CAR.codec.link(data)
+  const nb = { link, size: data.byteLength }
+  const can = 'store/add'
+
+  const request = await CAR.encode([
+    {
+      issuer: alice,
+      audience: uploadService,
+      capabilities: [{
+        can,
+        with: spaceDid,
+        nb
+      }],
+      proofs: [proof],
+    }
+  ])
+
+  // @ts-expect-error different type interface in AWS expected request
+  await persistUcanInvocation(request, ucanStore)
+
+  const requestCar = await CAR.codec.decode(request.body)
+  const requestCarRootCid = requestCar.roots[0].cid.toString()
+
+  const cmd = new GetObjectCommand({
+    Key: `${requestCarRootCid}/${requestCarRootCid}.car`,
+    Bucket: bucketName,
+  })
+  const s3Response = await t.context.s3Client.send(cmd)
+  t.is(s3Response.$metadata.httpStatusCode, 200)
+
+  // @ts-expect-error AWS types with readable stream
+  const bytes = (await s3Response.Body.toArray())[0]
```

Comment:
What is this wild magic? Why would a toArray method stick all the bytes in the first element of an array

Reply:
https://nodejs.org/api/stream.html#readabletoarrayoptions

It is kind of the easier way of obtaining the contents of a stream. And for a test purpose seems good